embedded/src/main/java/org/wildfly/core/embedded/EmbeddedHostControllerFactory.java

```diff
@@ -0,0 +1,462 @@
+ * </p>
+ * <p>
+ * To use this class the <code>jboss.home.dir</code> system property must be set to the
+ * application server home directory. By default it will use the directories
+ * <code>{$jboss.home.dir}/standalone/config</code> as the <i>configuration</i> directory and
+ * <code>{$jboss.home.dir}/standalone/data</code> as the <i>data</i> directory. This can be overridden
+ * with the <code>${jboss.server.base.dir}</code>, <code>${jboss.server.config.dir}</code> or <code>${jboss.server.config.dir}</code>
+ * system properties as for normal server startup.
+ * </p>
+ * <p>
+ * If a clean run is wanted, you can specify <code>${jboss.embedded.root}</code> to an existing directory
+ * which will copy the contents of the data and configuration directories under a temporary folder. This
+ * has the effect of this run not polluting later runs of the embedded server.
+ * </p>
+ *
+ * @author Ken Wills <kwills@redhat.com>
+ * @see EmbeddedServerFactory
+ */
+public class EmbeddedHostControllerFactory {
+
+    public static final String JBOSS_EMBEDDED_ROOT = "jboss.embedded.root";
+
+    private EmbeddedHostControllerFactory() {
+    }
+
+    public static HostController create(final File jbossHomeDir, final ModuleLoader moduleLoader, final Properties systemProps, final Map<String, String> systemEnv, final String[] cmdargs) {
+        if (jbossHomeDir == null) {
+            throw EmbeddedLogger.ROOT_LOGGER.nullVar("jbossHomeDir");
+        }
+        if (moduleLoader == null) {
+            throw EmbeddedLogger.ROOT_LOGGER.nullVar("moduleLoader");
+        }
+        if (systemProps == null) {
+            throw EmbeddedLogger.ROOT_LOGGER.nullVar("systemProps");
+        }
+        if (systemEnv == null) {
+            throw EmbeddedLogger.ROOT_LOGGER.nullVar("systemEnv");
+        }
+        if (cmdargs == null) {
+            throw EmbeddedLogger.ROOT_LOGGER.nullVar("cmdargs");
+        }
+
+        setupCleanDirectories(jbossHomeDir, systemProps);
+        return new HostControllerImpl(jbossHomeDir, cmdargs, systemProps, systemEnv, moduleLoader);
+    }
+
+    static void setupCleanDirectories(File jbossHomeDir, Properties props) {
+        File tempRoot = getTempRoot(props);
+        if (tempRoot == null) {
+            return;
+        }
+
+        File originalConfigDir = getFileUnderAsRoot(jbossHomeDir, props, HostControllerEnvironment.DOMAIN_CONFIG_DIR, "configuration", true);
+        File originalDataDir = getFileUnderAsRoot(jbossHomeDir, props, HostControllerEnvironment.DOMAIN_DATA_DIR, "data", false);
+
+        try {
+            File configDir = new File(tempRoot, "config");
+            Files.createDirectory(configDir.toPath());
+            File dataDir = new File(tempRoot, "data");
+            Files.createDirectory(dataDir.toPath());
+            // For jboss.server.deployment.scanner.default
+            File deploymentsDir = new File(tempRoot, "deployments");
+            Files.createDirectory(deploymentsDir.toPath());
```

Comment:
Why not use Path api all the way trough? if noting else it has much much better error reporting than File api.